feat: Ephemeral peer channels

Septias opened the pull request:
TLDR: Webxdc API to send ephemeral messages with iroh.

Desktop PR: ?
Documentation PR: ?
Supersedes: #5041

### WebXDC Ephemeral Channels API

The WebXDC API for developers will have two new methods `sendEphemeral(payload: Payload)` and `setEphemeralListener((payload: Payload) => void)` which allow them to explicitly send and receive data over an "ephemeral" channel. This data only has to be serializable by js. 

Ephemeral channels provide a fast transport layer, ideally directly between machines collaborating on the same WebXDC app, improving performance significantly. The downside of ephemeral messages is that they do _not_  persist like regular, mail-based WebXDC messages. WebXDC application developers need to explicitly handle persistence if desired (for example by sending regular messages via SMTP), but ideally only use ephemeral channels for non-critical data, like cursor positions.

### Iroh

DeltaChat wants to use the ephemeral-channel provider [Iroh](https://iroh.computer/), which attempts to establish direct p2p connections between peers (QUIC), using PlumTree as a gossiping protocol for efficient peer sampling. Iroh falls back to relay nodes when a direct connection can not be established. Even a relayed connection is faster than sending single WebXDC messages via mail. 

### Implementation
Ephemeral channels should be established lazily, to avoid bootstrapping p2p connectivity when it's not required. By asking application developers to explicitly subscribe to ephemeral messages, we know when this should happen.

In any case, we need a way to agree upon a gossip topic ID among all (future) participants, this topic ID should never be guessable or readable to any outsider.

1. We introduce a new `GossipTopic` message header with a random 32-byte TopicId, securely generated on the initial WebXDC sender's device. This message header is encrypted and sent in the same message as the WebXDC application.
2. Whenever  `setEphemeralListener` is called by the WebXDC application, we start a routine to establish p2p connectivity and join the gossip swarm with Iroh.
3. The first step of this routine is to introduce yourself with a regular message containing the `IrohPublicKey`. We need to make sure it does not contain any other information, like IP address as this information can be persisted by email providers.
4. After the announcement, the sending peer joins the gossip swarm with an empty list of peer IDs (as they don't know anyone yet).
5. Upon receiving an announcement message, other peers store the sender's NodeAddr in the database (scoped per WebXDC app instance/message-id). The other peers can then equally join the gossip with `setEphemeralListener`.
6. Only after receiving a pong from the other side, `sendEphemeral` messages have a chance to be delivered to anyone.

### Caveats / Open Questions
- When a user gets removed from a chat, they will still be able to receive messages on the ephemeral channel as they've learned about the TopicId.
- When the app is closed, we need a way to leave the swarm, currently, there's no way to determine if a WebXDC app was closed


Thanks to @adzialocha for writing up the discussed flow. I made a few changes based on the implementation so please check if the current description still suits you.

Simon-Laux commented on src/context.rs:
```diff
@@ -473,11 +483,18 @@ impl Context {
             }
         }
+
+        if let Err(e) = self.create_gossip().await {
+            warn!(self, "{e}");
+        }
```
- [ ] This is blocking, it delays my desktop startup time by 6 seconds with my 15 accounts. it should not block
- [ ] Also it might be better to do this when it's first used to risk to trigger any unexpected (to the user) firewall system dialogs at delta chat startup. (I don't know enough about iroh if it makes system calls that trigger this with `create_gossip` or not)

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,430 @@
+                    .join("iroh_gossip_peers")
+                    .to_path_buf(),
+            )
+            .relay_mode(RelayMode::Default)
```
What do want to do differently?

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    /// Get the topic for given [MsgId].
+    pub async fn get_topic_for_msg_id(&self, msg_id: MsgId) -> Result<TopicId> {
```
This is used in json-rpc, but only for a minor minor performance improvement. I can refactor this to be private.

Septias commented on deltachat-jsonrpc/src/api.rs:
```diff
@@ -1719,6 +1719,37 @@ impl CommandApi {
+        &self,
+        account_id: u32,
+        instance_msg_id: u32,
+        update_str: String,
```
Livetime problem because of macro usage.

Septias commented on deltachat-jsonrpc/src/lib.rs:
```diff
@@ -1,3 +1,4 @@
+#![recursion_limit = "256"]
```
Yes, otherwise compilation fails.

Septias commented on flake.nix:
```diff
@@ -11,7 +11,12 @@
-        pkgs = nixpkgs.legacyPackages.${system};
+        pkgs = (import nixpkgs {
+          inherit system;
+          overlays = [
```
Migration problem, will take this out later

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,430 @@
+        let endpoint = MagicEndpoint::builder()
+            .secret_key(secret_key)
+            .alpns(vec![GOSSIP_ALPN.to_vec()])
+            .peers_data_path(
```
We can store peer information in the database table `iroh_gossip_peers` if we don't want to use iroh internalt solution. File corruption is probably not a problem.

Septias commented on src/receive_imf.rs:
```diff
@@ -2296,7 +2340,7 @@ fn compute_mailinglist_name(
     {
         if let Some(display_name) = &mime_parser.from.display_name {
-            name = display_name.clone();
+            name.clone_from(display_name);
```
okay, will do

Septias commented:
Sending binary data in Base64 encoded strings is possible and should not produce too much overhead in comparison to network letency

Septias commented:
How do I fix cargo-deny problem about duplicate crate-version?

Septias commented:
> As long as one user is using a chatmail server, there is a DERP server everyone can connect to and otherwise it is unlikely to have connectivity anyway.

What do we do if no peer is part of a chatmail server? What is your proposed solution? Upon sending, the sender checks the group for the first chatmail address (how should one determine this?) and otherwise resorts to our own chatmail instance.

Septias commented on src/mimefactory.rs:
```diff
@@ -1303,6 +1314,22 @@ impl<'a> MimeFactory<'a> {
             let json = self.msg.param.get(Param::Arg).unwrap_or_default();
             parts.push(context.build_status_update_part(json));
         } else if self.msg.viewtype == Viewtype::Webxdc {
+            let topic = create_random_topic();
```
Sending any webxdc will boot up iroh machinery.

Septias commented:
Gossip API is currently only exposed to json-rpc

Septias commented:
Somehow clippy crashes on me due to overflow, can someone else check?

hpk42 commented:
even after "cargo clean" and "rm -rf target" i get the attached compile-error on jsonrpc. 

[crash.log](https://github.com/deltachat/deltachat-core-rust/files/15113330/crash.log)

hpk42 commented:
thanks @link2xt for making it compile better but there still is a new failure. 
Not sure if desktop or core is at fault but i get:

```
> deltachat-desktop@1.44.1 build:backend
> tsc --build src/main

src/main/deltachat/webxdc.ts:570:22 - error TS2339: Property 'sendWebxdcGossipAdvertisement' does not exist on type 'RawClient'.

570       await this.rpc.sendWebxdcGossipAdvertisement(accountId, msgId)
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Septias commented:
Test fails in merge commit to main. This could be anything...

iequidoo commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,611 @@
+
+impl Context {
+    /// Create magic endpoint and gossip for the context.
+    pub(crate) async fn inite_peer_channels(&self) -> Result<()> {
```
typo: init

iequidoo commented on src/mimefactory.rs:
```diff
@@ -1314,6 +1325,17 @@ impl<'a> MimeFactory<'a> {
             parts.push(context.build_status_update_part(json));
         } else if self.msg.viewtype == Viewtype::Webxdc {
+            let topic = create_random_topic();
+            let peer = Box::pin(context.get_or_generate_iroh_keypair())
```
Isn't this expensive if the app doesn't use iroh channels?

Septias replied:
This only happens only for webxdc messages which you don't send too often. Sending the 30kb should deminish the performance implication of two SQL-queries and creating some random bytes.

iequidoo commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,611 @@
+    /// Create magic endpoint and gossip for the context.
+    pub(crate) async fn inite_peer_channels(&self) -> Result<()> {
+        let secret_key: SecretKey = self.get_or_generate_iroh_keypair().await?;
+        info!(self, "Iroh secret key: {}", secret_key.to_string());
```
Recently all secrets like SecureJoin QRs were removed from the log, better not to add new ones

iequidoo commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,611 @@
+                    // This should later be RelayMode::Disable as soon as chatmail servers have relay servers
+                    .unwrap_or(RelayMode::Default), */
+                    RelayMode::Custom(RelayMap::from_url(RelayUrl::from(
+                        Url::parse("https://iroh.testrun.org:4443").unwrap(),
```
Better to make it the default value of a new config key so that it can be reconfigured via environment

Septias replied:
Gonna care about it when we are there. If its fine with you I'm resolving this for now..

Septias commented on src/context.rs:
```diff
@@ -288,6 +291,15 @@ pub struct InnerContext {
     pub(crate) push_subscribed: AtomicBool,
+
+    /// [MagicEndpoint] needed for iroh peer channels.
+    pub(crate) endpoint: Mutex<Option<MagicEndpoint>>,
```
Where do you expect deadlock?

iequidoo replied:
I see that `iroh_endpoint.write()` is called under `iroh_gossip.write()`. Also `iroh_channels.write()` is called under `iroh_gossip.read()`. Now there are no ABBA deadlocks, but it should be documented in which order mutexes/rwlocks must be acquired if it's really necessary to acquire multiple locks at the same time (i haven't read all the code yet, so haven't yet understood all this locking magic)

Septias replied:
Is it mathematically proven acquiring locks in the same order is sound?

iequidoo replied:
If you're about Rust, then there's no such restriction and that's one of the reasons why it doesn't protect from deadlocks. Taking locks in the same order everywhere is one of the ways to protect yourself (not the only of course, i just don't remeber them all). But maybe the code just can be refactored so that we never take multiple locks at the same time and don't need to document anything about locking?

Septias replied:
At the moment the code is not really functional, taking locks at multiple places, so a simple refactoring is not possible. We are currently trying to get this PR into main for the next release, so I would suggest not to do this kind of refactoring now, as there are a lot of other things needed and my capacities are limited at the moment.

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,610 @@
+    async fn _delete_webxdc_gossip_peer_for_msg(&self, topic: TopicId, peer: NodeId) -> Result<()> {
+        self.sql
+            .execute(
+                "DELETE FROM iroh_gossip_peers WHERE public_key = ? topic = ?",
```
we can remove it

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,610 @@
+        Ok(NodeAddr::from_parts(
+            endpoint
+                .as_ref()
+                .context("iroh endpoint not initialized")?
```
"value does not live long enough"

iequidoo commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,591 @@
+    }
+
+    /// Leave the gossip of the webxdc with given [MsgId].
+    pub async fn leave_realtime(&self, msg_id: MsgId) -> Result<()> {
```
Should be renamed to `leave_webxdc_realtime()` as well

iequidoo commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,591 @@
+    ///
+    /// The returned future resolves when the swarm becomes operational.
+    async fn join_and_subscribe_gossip(&self, msg_id: MsgId) -> Result<JoinTopicFut> {
+        let mut gossip = (*self.iroh_gossip.write().await).clone();
```
Why not `.read()`?

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,590 @@
+        ))
+    }
+
+    /// Get the topic for a given [MsgId].
+    pub(crate) async fn get_topic_for_msg_id(&self, msg_id: MsgId) -> Result<TopicId> {
```
I feel like MsgId shoudl just be some wrapper around the internal u32, I don't like it to have many special methods. I never know where to look for my methods.

iequidoo commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,593 @@
+//!    we start a routine to establish p2p connectivity and join the gossip swarm with Iroh.
+//! 3. The first step of this routine is to introduce yourself with a regular message containing the `IrohPublicKey`.
+//!    This message contains the users relay-server and public key.
+//!    Direct IP address are not included as this information can be persisted by email providers.
```
```suggestion
//!    Direct IP address is not included as this information can be persisted by email providers.
```

iequidoo commented on src/mimefactory.rs:
```diff
@@ -1314,6 +1325,17 @@ impl<'a> MimeFactory<'a> {
+            context
+                .iroh_add_peer_for_topic(self.msg.id, topic, peer, None)
+                .await?;
+            headers.protected.push(Header::new(
```
Is it guaranteed that the message is encrypted? `IrohGossipTopic` is a secret afaiu

hpk42 replied:
I don't think we need to worry about encryption here.  If we want to prevent sending an Iroh-announcement via unencrypted mails we should  do it more generically and not spawn/instantiate anything.  For now, i think it's fine to allow announces via unencrypted mails.

Septias replied:
It is also put into the protected headers (part of the encrypted part) so should be fine, no? @link2xt

Septias commented on src/qr.rs:
```diff
@@ -116,7 +116,7 @@ pub enum Qr {
         /// information to connect to and authenticate a backup provider.
         ///
         /// The format is somewhat opaque, but `sendme` can deserialise this.
-        ticket: iroh::provider::Ticket,
```
I actually like the clear distinction between old and new iroh here. We anyway plan to deprecate old iroh in the next months so this will not stay for long.

Septias commented on src/context.rs:
```diff
@@ -288,6 +291,17 @@ pub struct InnerContext {
+    pub(crate) iroh_gossip: RwLock<Option<Gossip>>,
+
+    /// Topics for which an advertisement has already been sent.
+    /// If iroh_channels lock is also needed, it should be aquired first.
```
The order you described is also what I deduced, the comment is misplaced indeed.

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,601 @@
+    /// Create magic endpoint and gossip for the context.
+    pub(crate) async fn init_peer_channels(&self) -> Result<()> {
+        let secret_key: SecretKey = self.get_or_generate_iroh_keypair().await?;
+        info!(self, "Secret key: {}", secret_key.public().to_string());
```
This was just for syncing the code to my laptop. I need two devices to test things.

dignifiedquire commented on src/context.rs:
```diff
@@ -288,6 +291,17 @@ pub struct InnerContext {
+
+    /// [MagicEndpoint] needed for iroh peer channels.
+    /// If iroh_gossip lock is also needed, it should be aquired first.
+    pub(crate) iroh_endpoint: RwLock<Option<MagicEndpoint>>,
```
A lot of the code could be simplified by using sth like https://docs.rs/async-once-cell/latest/async_once_cell/struct.OnceCell.html to do the lazy init, rather than using `RwLock<Option>`

Septias replied:
True, I don't know how I missed that..

dignifiedquire commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,597 @@
+                .context("iroh endpoint not initialized")?
+                .node_id(),
+            relay,
+            vec![],
```
Could just use `endpoint.my_addr()` and then delete the direct addresses, if you don't want them

dignifiedquire commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,597 @@
+        // depending on architecture this is 4 or 8 bytes.
+        // on some embedded deviced might be even more because of usize nature
+        let seq_num = has_joined.unwrap_or_default();
+        data.extend(seq_num.to_ne_bytes());
```
why use native endian?

Septias replied:
no particular reason, this is just to duplicate anyway.

dignifiedquire replied:
then I would recommend using either big or little endian fixed, for cross machine compatability

dignifiedquire commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,597 @@
+
+        // spawn endpoint loop that forwards incoming connections to the gossiper
+        let context = self.clone();
+        tokio::spawn(endpoint_loop(context, endpoint.clone(), gossip.clone()));
```
this task is not managed at all (shutdown etc)

Septias replied:
We want this to be shut down with delta chat, so this should be fine. Recovery might be needed, though...
Does this loop fail sometimes but is able to be reestablished?

dignifiedquire commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,597 @@
+            .join(topic, peers.into_iter().map(|addr| addr.node_id).collect())
+            .await?;
+
+        tokio::spawn(subscribe_loop(self.clone(), gossip.clone(), topic, msg_id));
```
this task is not managed at all (shutdown etc)

dignifiedquire commented on Cargo.toml:
```diff
@@ -59,8 +71,21 @@ futures-lite = "2.3.0"
+  "bmp",
+] }
+iroh_old = { package = "iroh", version = "0.4.2", default-features = false }
+iroh = { version = "0.15.0", default-features = false }
```
you shouldn't need this, given you are using the individual iroh crates directly

Septias commented:
I made a big refactoring and moved all iroh-related stuff behind one OnceCell one the context object. This way need no more locks and thus remove the threat of running into a deadlock. Also, the context is less poluted because I moved out of the context into top-level functions in the `peer_channels` module. This should lead to cleaner code.

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,688 @@
+    ///
+    /// If there is no gossip, create it.
+    ///
+    /// The returned future resolves when the swarm becomes operational.
```
Needed for testing

Septias commented on deltachat-jsonrpc/src/api.rs:
```diff
@@ -1730,6 +1733,37 @@ impl CommandApi {
+        &self,
+        account_id: u32,
+        instance_msg_id: u32,
+        data: Vec<u8>,
```
need a vec eventually and don't want to do the transformation

Septias commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,688 @@
+        let seq_num = self.get_and_incr(&topic).await;
+        data.extend(seq_num.to_le_bytes());
+
+        self.gossip.broadcast(topic, data.into()).await?;
```
actually, we receive a byte vector instead of base64 from the frontend

Septias commented on flake.nix:
```diff
@@ -525,15 +525,25 @@
           };
 
-        devShells.default = pkgs.mkShell {
+        devShells.default = let 
```
same as https://github.com/deltachat/deltachat-core-rust/pull/5548.
Should vanish when we merge the PR and the merge main.

dignifiedquire commented:
🎉

Hocuri commented on src/peer_channels.rs:
```diff
@@ -0,0 +1,711 @@
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn test_can_reconnect() {
```
This test is timing out in https://github.com/deltachat/deltachat-core-rust/pull/5619 in `Rust CI / Rust tests (ubuntu-latest, 1.77.0)`
